plugins/meta/iplink/main.go

```diff
@@ -0,0 +1,223 @@
+		return fmt.Errorf("Invalid args %v for MAC addr: %v", newMacAddr, err)
+	}
+
+	link, err := netlink.LinkByName(ifName)
+	if err != nil {
+		return fmt.Errorf("Failed to get %s: %v", ifName, err)
+	}
+
+	err = netlink.LinkSetDown(link)
+	if err != nil {
+		return fmt.Errorf("Failed to set %s down: %v", ifName, err)
+	}
+	err = netlink.LinkSetHardwareAddr(link, addr)
+	if err != nil {
+		return fmt.Errorf("Failed to set %s address to %s: %v", ifName, newMacAddr, err)
+	}
+	return netlink.LinkSetUp(link)
+}
+
+func updateResultsMacAddr(config PluginConf, ifName string, newMacAddr string) error {
+	for _, i := range config.PrevResult.Interfaces {
+		if i.Name == ifName {
+			i.Mac = newMacAddr
+		}
+	}
+	return nil
+}
+
+func changePromisc(ifName string, val bool) error {
+	link, err := netlink.LinkByName(ifName)
+	if err != nil {
+		return fmt.Errorf("Failed to get %s: %v", ifName, err)
+	}
+
+	if val {
+		return netlink.SetPromiscOn(link)
+	}
+	return netlink.SetPromiscOff(link)
+}
+
+func changeMtu(ifName string, mtu int) error {
+	link, err := netlink.LinkByName(ifName)
+	if err != nil {
+		return fmt.Errorf("Failed to get %s: %v", ifName, err)
+	}
+	return netlink.LinkSetMTU(link, mtu)
+}
+
+func parseConf(data []byte, envArgs string) (*PluginConf, error) {
+	conf := PluginConf{IpLinkConf: &IpLinkConf{Promisc: false, Mtu: -1}}
+	if err := json.Unmarshal(data, &conf); err != nil {
+		return nil, fmt.Errorf("failed to load netconf: %v", err)
+	}
+
+	// Parse custom MAC from both env args
+	if envArgs != "" {
+		e := MACEnvArgs{}
+		err := types.LoadArgs(envArgs, &e)
+		if err != nil {
+			return nil, err
+		}
+
+		if e.MAC != "" {
+			conf.Mac = string(e.MAC)
+		}
+	}
+
+	if conf.IpLinkConf == nil && conf.RuntimeConfig.IpLinkConf != nil {
+		conf.IpLinkConf = conf.RuntimeConfig.IpLinkConf
+	}
+
+	// Parse previous result.
+	if conf.RawPrevResult != nil {
+		resultBytes, err := json.Marshal(conf.RawPrevResult)
+		if err != nil {
+			return nil, fmt.Errorf("could not serialize prevResult: %v", err)
+		}
+		res, err := version.NewResult(conf.CNIVersion, resultBytes)
+		if err != nil {
+			return nil, fmt.Errorf("could not parse prevResult: %v", err)
+		}
+		conf.RawPrevResult = nil
+		conf.PrevResult, err = current.NewResultFromResult(res)
+		if err != nil {
+			return nil, fmt.Errorf("could not convert result to current version: %v", err)
+		}
+	}
+
+	return &conf, nil
+}
+
+func cmdAdd(args *skel.CmdArgs) error {
+	conf, err := parseConf(args.StdinData, args.Args)
+	if err != nil {
+		return err
+	}
+
+	err = ns.WithNetNSPath(args.Netns, func(_ ns.NetNS) error {
+		var err error
+
+		if conf.Mac != "" {
+			if err = changeMacAddr(args.IfName, conf.Mac); err == nil {
+				err = updateResultsMacAddr(*conf, args.IfName, conf.Mac)
+			}
+		}
+
+		if conf.Promisc != false {
+			if err = changePromisc(args.IfName, true); err != nil {
+				return err
+			}
+		}
+
+		if conf.Mtu != -1 {
+			if err = changeMtu(args.IfName, conf.Mtu); err != nil {
```

Comment:
MTU modified in this place is not updated in returned `Result`. It should be updated in similar way as hwaddr is updated.

Reply:
Currently there is no 'mtu' in CNI specification, https://github.com/containernetworking/cni/blob/master/SPEC.md#result, hence no need to update.

Adding 'mtu' into result should be taken care of another issues/PR, I suppose.

Reply:
Ah, you are right.